### PR TITLE
Pull request for the kvm-qemu-2.4.0.1-libvmi.patch, with updated README

### PR DESCRIPTION
**Pull request by valerioa**
This is an updated pull request for the  kvm-qemu-2.4.0.1-libvmi.patch. I've updated the README in the directory tools/

### CODE REVIEW

**Comment by namidairo** [comment]
You spelled length as lenght in the pmemaccess description. Also, test against guests with >2GB ram as I suspect it might have problems referenced in #258 (It was only fixed against Xen to my knowledge)

**Comment by valerioa** [comment]
Thanks, I'll correct the spelling mistakes. On the issue of VMs having more than 2GB of RAM, it seems not to affect QEMU/KVM

I have WindwosXP VM with 4GB RAM:

libvmi-0.10.1 # /usr/local/pkg/libvmi-0.10.1/bin/process-list WindowsXP
Max size is 4294967296
Process listing for VM WindowsXP (id=83)
Next list entry is at: 8a0d4c98
[    4] System (struct addr:8a3389c8)
[  548] smss.exe (struct addr:8a0d4c10)
[  612] csrss.exe (struct addr:8a0d0900)
[  636] winlogon.exe (struct addr:8a2018a8)
[  680] services.exe (struct addr:8a0e8da0)
[  692] lsass.exe (struct addr:8a0e17f8)
[  860] svchost.exe (struct addr:8a0efc10)
[  948] svchost.exe (struct addr:89ef9548)
[ 1044] svchost.exe (struct addr:89f0f5a8)
[ 1196] svchost.exe (struct addr:89e6b7a8)
[ 1264] svchost.exe (struct addr:89e7c7a8)
[ 1484] spoolsv.exe (struct addr:89e42020)
[ 1700] explorer.exe (struct addr:89e37958)
[ 1820] ctfmon.exe (struct addr:8a122650)
[ 1972] svchost.exe (struct addr:8a125da0)
[ 1244] alg.exe (struct addr:8a0e8020)
[ 1628] wscntfy.exe (struct addr:89ddf1c0)
[  828] wuauclt.exe (struct addr:8a196790)

'Max size is 4294967296' is a printf I added to show the size of the guest RAM.

or, maybe I need a description on how to reproduce the guest with > 2GB RAM problem

**Comment by awsaba** [comment]
The issue appears with Xen under Windows 7 x64, which will map virtual addresses to physical addresses > 4GB because of the way devices and other holes are mapped below 4GB.  I haven't looked into QEMU's pmemaccess enough to know how any holes are handled.

**Comment by tklengyel** [comment]
What's the status on this PR? If it's good to go, please squash the commits in this PR into a single one. Thanks!

File: tools/qemu-kvm-patch/kvm-qemu-2.4.0.1-libvmi.patch [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+diff -Naur qemu-2.4.0.1/qmp-commands.hx qemu-2.4.0.1-valerio/qmp-commands.hx
+--- qemu-2.4.0.1/qmp-commands.hx	2015-09-22 16:00:51.000000000 -0600
++++ qemu-2.4.0.1-valerio/qmp-commands.hx	2015-10-12 10:37:43.791802959 -0600
+@@ -467,6 +467,29 @@
+ EQMP
+ 
+     {
++        .name       = "pmemaccess",
++        .args_type  = "path:s",
++        .mhandler.cmd_new = qmp_marshal_input_pmemaccess,
```
Comment (namidairo) [inline tools/qemu-kvm-patch/kvm-qemu-2.4.0.1-libvmi.patch]: For anyone wondering about QEMU 2.5,

When you try to build this against QEMU 2.5.1 it'll end up looking for qmp_marshal_pmemaccess instead.

I'm guessing they derive the names from the schema automatically or something.

File: tools/qemu-kvm-patch/kvm-qemu-2.4.0.1-libvmi.patch [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
++ * Original Author: Bryan D. Payne (bdpayne@acm.org)
++ * 
++ * Refurbished for modern QEMU by Valerio Aimale (valerio@aimale.com), in 2015
++ */
++
++#include "memory-access.h"
++//#include "cpu-all.h"
++#include "qemu-common.h"
++#include "exec/cpu-common.h"
++#include "config.h"
```
Comment (namidairo) [inline tools/qemu-kvm-patch/kvm-qemu-2.4.0.1-libvmi.patch]: Would rather include `qemu/osdep.h` for compatibility with 2.6+, as config.h has been removed recently.

I believe I might be getting the same issue as #333 when I interrupt any of the examples and try to run another though. Seems to create another socket but libvmi not being able to see it for whatever reason. Perhaps a race condition, will have to test further.

File: tools/qemu-kvm-patch/kvm-qemu-2.4.0.1-libvmi.patch [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
++    }
++
++    connection_fd = accept(socket_fd, (struct sockaddr *) &address, &address_length);
++    connection_handler(connection_fd);
++
++    close(socket_fd);
++    unlink(pargs->path);
++    free(pargs->path);
++    free(pargs);
++error_exit:
```
Comment (ArtemKotovsky) [inline tools/qemu-kvm-patch/kvm-qemu-2.4.0.1-libvmi.patch]: need move above, memory leak on error

File: tools/qemu-kvm-patch/kvm-qemu-2.4.0.1-libvmi.patch [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
++
++    if (bind(socket_fd, (struct sockaddr *) &address, address_length) != 0){
++	error_setg(pargs->errp, "Qemu pmemaccess: bind failed");
++        goto error_exit;
++    }
++    if (listen(socket_fd, 0) != 0){
++	error_setg(pargs->errp, "Qemu pmemaccess: listen failed");
++        goto error_exit;
++    }
++
```
Comment (ArtemKotovsky) [inline tools/qemu-kvm-patch/kvm-qemu-2.4.0.1-libvmi.patch]: qmp_pmemaccess must wait until the thread init resources, sometimes, LibVMI can't open socket with error=2

File: tools/qemu-kvm-patch/kvm-qemu-2.4.0.1-libvmi.patch [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
++#ifndef MEMORY_ACCESS_H
++#define MEMORY_ACCESS_H
++
++#include "qapi-types.h"
++#include "qapi/qmp/qdict.h"
++#include "qapi/error.h"
++
++void qmp_pmemaccess (const char *path, Error **errp);
++
++struct pmemaccess_args {
```
Comment (ArtemKotovsky) [inline tools/qemu-kvm-patch/kvm-qemu-2.4.0.1-libvmi.patch]: can move to .c

File: tools/qemu-kvm-patch/kvm-qemu-2.4.0.1-libvmi.patch [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
++{
++    pthread_t thread;
++    sigset_t set, oldset;
++    struct pmemaccess_args *pargs;
++
++    // create the args struct
++    pargs = (struct pmemaccess_args *) malloc(sizeof(struct pmemaccess_args));
++    pargs->errp = errp;
++    // create a copy of path that we can safely use
++    pargs->path = malloc(strlen(path) + 1);
```
Comment (ArtemKotovsky) [inline tools/qemu-kvm-patch/kvm-qemu-2.4.0.1-libvmi.patch]: check malloc results

File: tools/qemu-kvm-patch/kvm-qemu-2.4.0.1-libvmi.patch [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
++    memcpy(pargs->path, path, strlen(path) + 1);
++	
++    // start the thread
++    sigfillset(&set);
++    pthread_sigmask(SIG_SETMASK, &set, &oldset);
++    pthread_create(&thread, NULL, memory_access_thread, pargs);
```
Comment (ArtemKotovsky) [inline tools/qemu-kvm-patch/kvm-qemu-2.4.0.1-libvmi.patch]: memory leak if thread hasn't started

**Comment by namidairo** [comment]
You may also need to check MemoryRegion->skip_dump like the dumper in QEMU does, since a read of the pci mmio spaces might crash the guest in some cases?

Eg. https://github.com/qemu/qemu/blob/stable-2.4/memory_mapping.c#L205-L209

Not sure how significant the overhead in creating a MemoryRegion object each time we do a read/write to do the comparison though.

**Comment by smaresca** [comment]
@valerioa thanks again for the PR. Have you had a chance to take a look at some of the feedback?

**Comment by valerioa** [comment]
@smaresca , I have seen the feedback. I think this patch need to be refurbished to be usable with the 2.6 ( and possibly 2.7) QEMU branch. I need to find some time to adapt it to the new releases.

**Comment by ptokponnon** [comment]
Can we have this patch for qemu 2.8?
the one available in the download is too obsolete...

**Comment by tklengyel** [comment]
@ptokponnon I don't think anyone is working on this right now so feel free to rebase this patch to 2.8 and open a PR for it!

**Pull request closed without merging**